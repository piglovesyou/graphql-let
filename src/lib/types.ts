export type CommandOpts = {
  cwd: string;
  configPath: string;
};

export type ConfigTypes = {
  generateDir: string;
  config?: Record<string, boolean | string>;
  schema: string;
  documents: string | string[];
  plugins: string[];
};
