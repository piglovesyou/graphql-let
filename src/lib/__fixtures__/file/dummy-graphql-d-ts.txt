/* a7870383395320c75dc806f677ffa7caad7522bc
 * This file is automatically generated by graphql-let.
 */
