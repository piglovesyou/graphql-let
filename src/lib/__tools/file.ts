import copyfiles from 'copyfiles';
import { promises } from 'fs';
import { join as pathJoin, sep } from 'path';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

const crlf = /\r\n/g;
export function normalizeNewLine(str: string) {
  if (crlf.test(str)) return str.replace(crlf, '\n');
  return str;
}

export const { readFile: _readFile, writeFile, rename } = promises;

export function readFile(file: string) {
  return _readFile(file, 'utf-8').then(normalizeNewLine);
}

export function cleanup(cwd: string, relPaths: string[]) {
  return Promise.all(relPaths.map((rel) => rimraf(pathJoin(cwd, rel))));
}

export function copyDir(
  baseFullDir: string,
  srcRelDir: string,
  destRelDir: string,
): Promise<void> {
  // baseFullDir: /Users/a/b/c
  // srcRelDir: d/e
  // destRelDir: .d/e
  // It generates /Users/a/b/c/.d/e
  if (srcRelDir === destRelDir) throw new Error('Kidding me?');
  const up = pathJoin(baseFullDir).split(sep).length + 1;
  const [relRoot] = destRelDir.split('/');
  return new Promise<void>((resolve, rejects) => {
    copyfiles(
      [pathJoin(baseFullDir, srcRelDir, '**'), pathJoin(baseFullDir, relRoot)],
      { error: true, up, all: true },
      (err) => {
        if (err) return rejects(err);
        resolve();
      },
    );
  });
}

export type AbsFn = (rel: string) => string;

export async function prepareFixtures(
  baseFullDir: string,
  fixtureSrcRelDir: string,
): Promise<[cwd: string, abs: AbsFn]> {
  if (fixtureSrcRelDir.startsWith('..'))
    throw new Error(
      `It doesn't support backward relative paths like ${fixtureSrcRelDir}`,
    );
  const fixtureDestRelDir = '.' + fixtureSrcRelDir;
  const cwd = pathJoin(baseFullDir, fixtureDestRelDir);
  const abs = (relPath: string) => pathJoin(cwd, relPath);
  await copyDir(baseFullDir, fixtureSrcRelDir, fixtureDestRelDir);
  return [cwd, abs];
}
