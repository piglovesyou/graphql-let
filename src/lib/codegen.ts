import { codegen as graphqlCodegen } from '@graphql-codegen/core';
import { promises as fsPromises } from 'fs';
import gql from 'graphql-tag';
import _mkdirp from 'mkdirp';
import path from 'path';
import { promisify } from 'util';
import genDts from './gen-dts';
import { PartialCodegenOpts } from './create-codegen-opts';
import { printInfo } from './print';
import { ConfigTypes } from './types';
import { existsSync } from 'fs';

const { readFile, writeFile } = fsPromises;
const mkdirp = promisify(_mkdirp);

// For the loader can process a same file simultaneously
const processingTasks = new Map<string /*fileFullPath*/, Promise<string>>();

function wrapAsModule(fileName: string, content: string) {
  return `declare module '*/${fileName}' {
  ${content.replace(/\n/g, '\n  ')}}`;
}

async function processGraphQLCodegen(
  codegenOpts: PartialCodegenOpts,
  tsxFullPath: string,
  gqlFullPath: string,
  gqlContent: string,
): Promise<string> {
  const tsxContent = await graphqlCodegen({
    ...codegenOpts,
    filename: tsxFullPath,
    documents: [
      {
        filePath: gqlFullPath,
        content: gql(gqlContent),
      },
    ],
  });
  await mkdirp(path.dirname(tsxFullPath));
  await writeFile(tsxFullPath, tsxContent);
  return tsxContent;
}

async function processGenDts(
  dtsFullPath: string,
  tsxFullPath: string,
  gqlFullPath: string,
  dtsRelPath: string,
) {
  await mkdirp(path.dirname(dtsFullPath));
  const dtsContent = await genDts(tsxFullPath);
  if (!dtsContent) throw new Error(`Generate ${dtsFullPath} fails.`);
  await writeFile(
    dtsFullPath,
    wrapAsModule(path.basename(gqlFullPath), dtsContent),
  );
  printInfo(`${dtsRelPath} was generated.`);
  return dtsContent;
}

export async function codegen(
  gqlContent: string,
  gqlFullPath: string,
  tsxFullPath: string,
  dtsRelPath: string,
  dtsFullPath: string,
  options: ConfigTypes,
  codegenOpts: PartialCodegenOpts,
): Promise<string> {
  let tsxContent: string;
  if (existsSync(tsxFullPath)) {
    tsxContent = await readFile(tsxFullPath, 'utf-8');
  } else if (processingTasks.has(tsxFullPath)) {
    tsxContent = await processingTasks.get(tsxFullPath)!;
  } else {
    const tsxPromise = processGraphQLCodegen(
      codegenOpts,
      tsxFullPath,
      gqlFullPath,
      gqlContent,
    );
    processingTasks.set(tsxFullPath, tsxPromise);
    tsxContent = await tsxPromise;
    processingTasks.delete(tsxFullPath);
  }

  if (existsSync(dtsFullPath) || processingTasks.has(dtsFullPath)) {
    // Already exists or is processing. Just skip.
  } else {
    const dtsPromise = processGenDts(
      dtsFullPath,
      tsxFullPath,
      gqlFullPath,
      dtsRelPath,
    );
    processingTasks.set(dtsFullPath, dtsPromise);
    await dtsPromise;
    processingTasks.delete(dtsFullPath);
  }

  return tsxContent;
}
