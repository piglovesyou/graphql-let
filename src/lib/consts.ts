export const DEFAULT_CONFIG_FILENAME = '.graphql-let.yml';
