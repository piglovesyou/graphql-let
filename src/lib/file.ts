import { promises as fsPromises } from 'fs';
import { createReadStream, existsSync } from 'fs';
import glob from 'globby';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

export const { readFile, writeFile } = fsPromises;
export { readFileSync, statSync } from 'fs';

// Erasing old cache in __generated__ on HMR.
// Otherwise the multiple `declare module "*/x.graphql"` are exposed.
export async function removeByPatterns(cwd: string, ...patterns: string[]) {
  const oldFiles = await glob(patterns, {
    cwd,
    absolute: true,
  });
  await Promise.all(oldFiles.map((f) => rimraf(f)));
}

const leadingStringOfGeneratedContent = '/* ';
const hexHashLength = 40;
export function withHash(sourceHash: string, content: string): string {
  return `${leadingStringOfGeneratedContent}${sourceHash}
 * This file is automatically generated by graphql-let. */

${content}`;
}

export function readHash(filePath: string): Promise<string | null> {
  if (!existsSync(filePath)) {
    return Promise.resolve(null);
  }

  return new Promise((resolve, reject) => {
    const stream = createReadStream(filePath, {
      encoding: 'utf-8',
      highWaterMark: leadingStringOfGeneratedContent.length + hexHashLength,
    });
    stream.on('error', (error) => reject(error));
    stream.on('data', (chunk: string) => {
      const hash = chunk.slice(leadingStringOfGeneratedContent.length);
      if (hash.length !== hexHashLength) return resolve(null);

      resolve(hash);
    });
    stream.read();
  });
}
