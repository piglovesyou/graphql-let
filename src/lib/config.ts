import { Types } from '@graphql-codegen/plugin-helpers';
import { join as pathJoin } from 'path';
import { parse as parseYaml } from 'yaml';
import { DEFAULT_CONFIG_FILENAME } from './consts';
import { readFile, readFileSync } from './file';
import { createHash } from './hash';
import { printError } from './print';

export type PartialGraphqlCodegenOptions = Omit<Types.Config, 'generates'>;

export type GraphQLLetAdditionalOptions = {
  plugins: Array<string | Record<string, any>>;
  respectGitIgnore?: boolean;
  schemaEntrypoint?: string;
  cacheDir?: string;
  TSConfigFile?: string;
  gqlDtsEntrypoint?: string;
  generateOptions?: Types.ConfiguredOutput;
};

export type UserConfigTypes = PartialGraphqlCodegenOptions &
  GraphQLLetAdditionalOptions;
export type ConfigTypes = PartialGraphqlCodegenOptions & {
  documents: Types.OperationDocumentGlobPath[];
} & Required<GraphQLLetAdditionalOptions>;

export function buildConfig(raw: UserConfigTypes): ConfigTypes {
  if (typeof raw !== 'object')
    printError(new Error('A config file must shape an object'));

  if (!raw.schema || !raw.documents || !raw.plugins)
    printError(new Error(`A config requires a "${name}" field`));

  const documents: Types.OperationDocumentGlobPath[] = Array.isArray(
    raw.documents,
  )
    ? (raw.documents as Types.OperationDocumentGlobPath[])
    : typeof raw.documents === 'string'
    ? [raw.documents]
    : (printError(
        new Error(`config.documents should be an array or a string`),
      ) as never);

  return {
    ...raw,
    // Normalized codegen options
    documents,
    // Set graphql-let default values
    respectGitIgnore: raw.respectGitIgnore || true,
    cacheDir: raw.cacheDir || 'node_modules/graphql-let/__generated__',
    TSConfigFile: raw.TSConfigFile || 'tsconfig.json',
    gqlDtsEntrypoint:
      raw.gqlDtsEntrypoint || 'node_modules/@types/graphql-let/index.d.ts',
    generateOptions: raw.generateOptions || Object.create(null),
    schemaEntrypoint: raw.schemaEntrypoint || '',
  };
}

export const getConfigPath = (cwd: string, configFilePath?: string) =>
  pathJoin(cwd, configFilePath || DEFAULT_CONFIG_FILENAME);

const getConfigFromContent = (content: string): [ConfigTypes, string] => [
  buildConfig(parseYaml(content)),
  createHash(content),
];

export default async function loadConfig(
  cwd: string,
  configFilePath?: string,
): Promise<[ConfigTypes, string]> {
  const configPath = getConfigPath(cwd, configFilePath);
  const content = await readFile(configPath, 'utf-8');
  return getConfigFromContent(content);
}

export function loadConfigSync(
  cwd: string,
  configFilePath?: string,
): [ConfigTypes, string] {
  const configPath = getConfigPath(cwd, configFilePath);
  const content = readFileSync(configPath, 'utf-8');
  return getConfigFromContent(content);
}
