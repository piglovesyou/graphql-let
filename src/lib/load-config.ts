import { join as pathJoin } from 'path';
import { parse as parseYaml } from 'yaml';
import { DEFAULT_CONFIG_FILENAME } from './consts';
import { readFile, readFileSync } from './file';
import { ConfigTypes } from './types';
import { createHash } from './hash';

const getConfigPath = (cwd: string, configFilePath?: string) =>
  pathJoin(cwd, configFilePath || DEFAULT_CONFIG_FILENAME);
const getConfigFromContent = (content: string): [ConfigTypes, string] => [
  parseYaml(content),
  createHash(content),
];

// TODO:
// function validate(config: ConfigTypes) { }
// function fillDefaultValues(config: ConfigTypes) { }

export default async function loadConfig(
  cwd: string,
  configFilePath?: string,
): Promise<[ConfigTypes, string]> {
  const configPath = getConfigPath(cwd, configFilePath);
  const content = await readFile(configPath, 'utf-8');
  return getConfigFromContent(content);
}

export function loadConfigSync(
  cwd: string,
  configFilePath?: string,
): [ConfigTypes, string] {
  const configPath = getConfigPath(cwd, configFilePath);
  const content = readFileSync(configPath, 'utf-8');
  return getConfigFromContent(content);
}
