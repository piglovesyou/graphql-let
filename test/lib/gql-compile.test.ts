import { deepStrictEqual, ok } from 'assert';
import { join as pathJoin } from 'path';
import { rimraf } from '../../src/lib/file';
import {
  GqlCodegenContext,
  processGqlCompile,
} from '../../src/lib/gql-compile';
import { buildConfig } from '../../src/lib/load-config';
import { ConfigTypes } from '../../src/lib/types';

const dtsRelDir = 'node_modules/@types/graphql-let';
const libRelDir = 'node_modules/graphql-let';

const cwd = pathJoin(__dirname, '../__fixtures/gql-compile');

const config: ConfigTypes = buildConfig({
  schema: 'schema/type-defs.graphqls',
  plugins: ['typescript', 'typescript-operations', 'typescript-react-apollo'],
  documents: [],
  respectGitIgnore: true,
  config: {
    reactApolloVersion: '3',
    withHOC: false,
    withHooks: true,
  },
});

describe('gql-compile', () => {
  beforeAll(async () => {
    await rimraf(pathJoin(cwd, 'node_modules'));
  });
  it(
    'compiles',
    async () => {
      const sourceRelPath = 'pages/index.tsx';
      const schemaHash = '234';
      const gqlContents = [
        `query Viewer {
    viewer {
        id
        name
        status
    }
}`,
      ];
      const codegenContext: GqlCodegenContext = [];
      const oldGqlContentHashes = new Set<string>();
      // const skippedContext: GqlCodegenContext = [];

      await processGqlCompile(
        cwd,
        config,
        dtsRelDir,
        pathJoin(libRelDir, '__generated__'),
        sourceRelPath,
        schemaHash,
        gqlContents,
        {},
        codegenContext,
        oldGqlContentHashes,
        // skippedContext,
      );

      deepStrictEqual(
        codegenContext[0].gqlContent,
        'query Viewer {\n    viewer {\n        id\n        name\n        status\n    }\n}',
      );
      deepStrictEqual(
        codegenContext[0].strippedGqlContent,
        'query Viewer{viewer{id name status}}',
      );
      deepStrictEqual(
        codegenContext[0].gqlContentHash,
        'dd28f9c0ad11900a2654540e86de9cf9fc16f8b4',
      );
      deepStrictEqual(codegenContext[0].sourceRelPath, 'pages/index.tsx');
      ok(
        codegenContext[0].sourceFullPath.endsWith(
          'graphql-let/test/__fixtures/gql-compile/pages/index.tsx',
        ),
      );
      deepStrictEqual(
        codegenContext[0].tsxRelPath,
        'pages/index-dd28f9c0ad11900a2654540e86de9cf9fc16f8b4.tsx',
      );
      ok(
        codegenContext[0].tsxFullPath.endsWith(
          'graphql-let/test/__fixtures/gql-compile/node_modules/graphql-let/__generated__/pages/index-dd28f9c0ad11900a2654540e86de9cf9fc16f8b4.tsx',
        ),
      );
      deepStrictEqual(
        codegenContext[0].dtsRelPath,
        'pages/index-dd28f9c0ad11900a2654540e86de9cf9fc16f8b4.d.ts',
      );
      ok(
        codegenContext[0].dtsFullPath.endsWith(
          'graphql-let/test/__fixtures/gql-compile/node_modules/@types/graphql-let/pages/index-dd28f9c0ad11900a2654540e86de9cf9fc16f8b4.d.ts',
        ),
      );
      deepStrictEqual(oldGqlContentHashes.size, 0);
    },
    1000 * 1000,
  );
});
