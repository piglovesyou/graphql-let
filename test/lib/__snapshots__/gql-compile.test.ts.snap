// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gql-compile compiles: /Users/soichitakamura/IdeaProjects/graphql-let/test/__fixtures/gql-compile/node_modules/@types/graphql-let/pages/index-dd28f9c0ad11900a2654540e86de9cf9fc16f8b4.d.ts 1`] = `
"/// <reference types=\\"react\\" />
import * as ApolloReactCommon from '@apollo/react-common';
import * as ApolloReactComponents from '@apollo/react-components';
import * as ApolloReactHooks from '@apollo/react-hooks';
export declare type Maybe<T> = T | null;
export declare type Exact<T extends {
  [key: string]: any;
}> = { [K in keyof T]: T[K] };
export declare type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
/** All built-in and custom scalars, mapped to their actual values */

export declare type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};
export declare type Query = {
  __typename?: 'Query';
  viewer?: Maybe<User>;
};
export declare type User = {
  __typename?: 'User';
  id: Scalars['ID'];
  name: Scalars['String'];
  status: Scalars['String'];
};
export declare type ViewerQueryVariables = Exact<{
  [key: string]: never;
}>;
export declare type ViewerQuery = ({
  __typename?: 'Query';
} & {
  viewer?: Maybe<({
    __typename?: 'User';
  } & Pick<User, 'id' | 'name' | 'status'>)>;
});
export declare const ViewerDocument: import(\\"graphql\\").DocumentNode;
export declare type ViewerComponentProps = Omit<ApolloReactComponents.QueryComponentOptions<ViewerQuery, ViewerQueryVariables>, 'query'>;
export declare const ViewerComponent: (props: ViewerComponentProps) => JSX.Element;
/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */

export declare function useViewerQuery(baseOptions?: ApolloReactHooks.QueryHookOptions<ViewerQuery, ViewerQueryVariables>): ApolloReactCommon.QueryResult<ViewerQuery, Exact<{
  [key: string]: never;
}>>;
export declare function useViewerLazyQuery(baseOptions?: ApolloReactHooks.LazyQueryHookOptions<ViewerQuery, ViewerQueryVariables>): ApolloReactHooks.QueryTuple<ViewerQuery, Exact<{
  [key: string]: never;
}>>;
export declare type ViewerQueryHookResult = ReturnType<typeof useViewerQuery>;
export declare type ViewerLazyQueryHookResult = ReturnType<typeof useViewerLazyQuery>;
export declare type ViewerQueryResult = ApolloReactCommon.QueryResult<ViewerQuery, ViewerQueryVariables>;
export declare type __AllExports = {
  Maybe: Maybe;
  Exact: Exact;
  Omit: Omit;
  Scalars: Scalars;
  Query: Query;
  User: User;
  ViewerQueryVariables: ViewerQueryVariables;
  ViewerQuery: ViewerQuery;
  ViewerDocument: ViewerDocument;
  ViewerComponentProps: ViewerComponentProps;
  ViewerComponent: ViewerComponent;
  useViewerQuery: typeof useViewerQuery;
  useViewerLazyQuery: typeof useViewerLazyQuery;
  ViewerQueryHookResult: ViewerQueryHookResult;
  ViewerLazyQueryHookResult: ViewerLazyQueryHookResult;
  ViewerQueryResult: ViewerQueryResult;
};"
`;

exports[`gql-compile compiles: /Users/soichitakamura/IdeaProjects/graphql-let/test/__fixtures/gql-compile/node_modules/graphql-let/__generated__/pages/index-dd28f9c0ad11900a2654540e86de9cf9fc16f8b4.tsx 1`] = `
"import gql from 'graphql-tag';
import * as React from 'react';
import * as ApolloReactCommon from '@apollo/react-common';
import * as ApolloReactComponents from '@apollo/react-components';
import * as ApolloReactHooks from '@apollo/react-hooks';
export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: any }> = { [K in keyof T]: T[K] };
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Query = {
  __typename?: 'Query';
  viewer?: Maybe<User>;
};

export type User = {
  __typename?: 'User';
  id: Scalars['ID'];
  name: Scalars['String'];
  status: Scalars['String'];
};

export type ViewerQueryVariables = Exact<{ [key: string]: never; }>;


export type ViewerQuery = (
  { __typename?: 'Query' }
  & { viewer?: Maybe<(
    { __typename?: 'User' }
    & Pick<User, 'id' | 'name' | 'status'>
  )> }
);


export const ViewerDocument = gql\`
    query Viewer {
  viewer {
    id
    name
    status
  }
}
    \`;
export type ViewerComponentProps = Omit<ApolloReactComponents.QueryComponentOptions<ViewerQuery, ViewerQueryVariables>, 'query'>;

    export const ViewerComponent = (props: ViewerComponentProps) => (
      <ApolloReactComponents.Query<ViewerQuery, ViewerQueryVariables> query={ViewerDocument} {...props} />
    );
    

/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */
export function useViewerQuery(baseOptions?: ApolloReactHooks.QueryHookOptions<ViewerQuery, ViewerQueryVariables>) {
        return ApolloReactHooks.useQuery<ViewerQuery, ViewerQueryVariables>(ViewerDocument, baseOptions);
      }
export function useViewerLazyQuery(baseOptions?: ApolloReactHooks.LazyQueryHookOptions<ViewerQuery, ViewerQueryVariables>) {
          return ApolloReactHooks.useLazyQuery<ViewerQuery, ViewerQueryVariables>(ViewerDocument, baseOptions);
        }
export type ViewerQueryHookResult = ReturnType<typeof useViewerQuery>;
export type ViewerLazyQueryHookResult = ReturnType<typeof useViewerLazyQuery>;
export type ViewerQueryResult = ApolloReactCommon.QueryResult<ViewerQuery, ViewerQueryVariables>;"
`;
